### ISSUE TITLE
module 'xgboost.compat' has no attribute 'XGBoostLabelEncoder' Error

### ISSUE THREAD

**Issue by gitshashwat** [dmlc/xgboost#6209]
When trying to use XGBoost, I am able to do everything but if I save the model using` joblib.dump(model, model_fn)` and try to load it back using `joblib.load('model.mdl')`, it is unable to load. It throws an error as **module 'xgboost.compat' has no attribute 'XGBoostLabelEncoder' Error**

XGBoost version: 0.9

**Comment by trivialfis** [comment]
Please upgrade to 1.2, and install sklearn.

**Comment by gitshashwat** [comment]
I've already installed sklearn. I'll consider upgrading to 1.2 too, but do we have any alternative?

**Comment by trivialfis** [comment]
I have never seen this error, actually I suspect that something went wrong with your installation or somehow `xgboost.compant` is changed to point to other entities.  Could you please look into whether `XGBoostLabelEncoder` is presented:

```
import xgboost
print(xgboost.compat.__file__)
print(xgboost.compat.XGBoostLabelEncoder)
```

**Comment by trivialfis** [comment]
Just checking.  Are you loading the pickle with same Python environment?  Like same version of xgboost and same version of Python?

**Comment by gitshashwat** [comment]
@trivialfis I was basically trying to load the model on a colab. I have saved in using my local system. And versions of Python and xgb is the same.

**Comment by trivialfis** [comment]
Do you have sklearn on colab?  Aside from that, it's advised to use `save_model` instead.  I know there was suggestion on using pkl on sklearn interface.   We changed that after 1.0 and `save_model` should be used all the time.  One of the reasons is pickle saves the whole environment instead of what matters (the model).  This doc is added in 1.0: https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html

**Comment by gitshashwat** [comment]
Cool, understood. This might be the real problem. The saving environment one. Keeping this open to comments for a week, and will close the issue otherwise.
Thanks @trivialfis :)

**Comment by trivialfis** [comment]
It a mix of many issues actually.  This involves the stability of Python byte code and the Python code structure.  Sklearn advice to use some model exporting libraries like onnx and gives a warning that some attributes might be lost in round-trip.  When using lossless pickle, it gives a warning that stability of pickle is not guaranteed.

On the other hand, without pickle it's really difficult to save information defined in Python.  In your case, XGBoost uses sklearn label encoder for preprocessing categorical labels, which is completely defined in Python and not controlled by XGBoost.  After 1.0, we save the encoder as a form of meta data in `save_model`, and attempt to reconstruct it during load.  But as you can see there will be edge cases.

cc @hcho3 My opinion is we remove the label encoder from xgboost completely and let users to define it themselves in their pipeline.  But that will be a huge breaking change.

**Comment by trivialfis** [comment]
Feel free to reopen if the issue is not yet solved.
